quaint/src/connector/ser.rs

```diff
@@ -0,0 +1,548 @@
+    where
+        S: Serializer,
+    {
+        let val = &self.0;
+
+        match &val.typed {
+            ValueType::Array(Some(values)) => {
+                let mut seq = serializer.serialize_seq(Some(values.len()))?;
+
+                for value in values {
+                    seq.serialize_element(&SerializedValue(value))?;
+                }
+
+                seq.end()
+            }
+            ValueType::Array(None) => serializer.serialize_none(),
+            ValueType::Int32(value) => value.serialize(serializer),
+            ValueType::Int64(value) => value.map(|val| val.to_string()).serialize(serializer),
+            ValueType::Numeric(value) => value
+                .as_ref()
+                .map(|value| value.normalized().to_string())
+                .serialize(serializer),
+            ValueType::Float(value) => value.serialize(serializer),
+            ValueType::Double(value) => value.serialize(serializer),
+            ValueType::Text(value) => value.serialize(serializer),
+            ValueType::Enum(value, _) => value.as_ref().map(|value| value.inner()).serialize(serializer),
+            ValueType::EnumArray(Some(variants), _) => {
+                let mut seq = serializer.serialize_seq(Some(variants.len()))?;
+
+                for variant in variants {
+                    seq.serialize_element(variant.inner())?;
+                }
+
+                seq.end()
+            }
+            ValueType::EnumArray(None, _) => serializer.serialize_none(),
+            ValueType::Bytes(value) => value.as_ref().map(base64::encode).serialize(serializer),
+            ValueType::Boolean(value) => value.serialize(serializer),
+            ValueType::Char(value) => value.serialize(serializer),
+            ValueType::Json(value) => value.serialize(serializer),
+            ValueType::Xml(value) => value.serialize(serializer),
+            ValueType::Uuid(value) => value.serialize(serializer),
+            ValueType::DateTime(value) => value.map(|value| value.to_rfc3339()).serialize(serializer),
+            ValueType::Date(value) => value.serialize(serializer),
+            ValueType::Time(value) => value.serialize(serializer),
+        }
+    }
+}
+
+#[derive(Debug, Copy, Clone, PartialEq, Serialize)]
+enum SerializedValueType {
+    #[serde(rename = "int")]
+    Int32,
+    #[serde(rename = "bigint")]
+    Int64,
+    #[serde(rename = "float")]
+    Float,
+    #[serde(rename = "double")]
+    Double,
+    #[serde(rename = "string")]
+    Text,
+    #[serde(rename = "enum")]
+    Enum,
+    #[serde(rename = "bytes")]
+    Bytes,
+    #[serde(rename = "bool")]
+    Boolean,
+    #[serde(rename = "char")]
+    Char,
+    #[serde(rename = "decimal")]
+    Numeric,
+    #[serde(rename = "json")]
+    Json,
+    #[serde(rename = "xml")]
+    Xml,
+    #[serde(rename = "uuid")]
+    Uuid,
+    #[serde(rename = "datetime")]
+    DateTime,
+    #[serde(rename = "date")]
+    Date,
+    #[serde(rename = "time")]
+    Time,
```

Comment:
Tried to clean that up using this macro:

```rs
macro_rules! impl_value_type {
    (
        $($variant:ident => ($mapped_variant:ident, $rename:expr)),*;
        $($array_variant:ident[] => ($mapped_array_variant:ident, $array_rename:expr)),*;
        $(($custom_variant:ident, $custom_rename:expr)),*
    ) => {
        #[derive(Debug, Clone, Serialize)]
        enum SerializedValueType {
            $(
                #[serde(rename = $rename)]
                $mapped_variant,
            )*
            $(
                #[serde(rename = $array_rename)]
                $mapped_array_variant,
            )*
            $(
                #[serde(rename = $custom_rename)]
                $custom_variant,
            )*
        }

        impl<'a> From<Value<'a>> for SerializedValueType {
            fn from(value: Value<'a>) -> Self {
                match &value.typed {
                    $(
                        ValueType::$variant(..) => SerializedValueType::$mapped_variant,
                    )*
                    ValueType::Array(values) => {
                        match values {
                            Some(values) if values.is_empty() => SerializedValueType::UnknownArray,
                            None => SerializedValueType::UnknownArray,
                            Some(values) => {
                                match &values[0].typed {
                                    $(
                                        ValueType::$array_variant(..) => SerializedValueType::$mapped_array_variant,
                                    )*
                                    ValueType::Array(_) | ValueType::EnumArray(_, _) => unreachable!()
                                }
                            }
                        }
                    }
                }
            }
        }
    }
}

impl_value_type! {
    Int32 => (Int32, "int"),
    Int64 => (Int64, "bigint"),
    Float => (Float, "float"),
    Double => (Double, "double"),
    Text => (String, "string"),
    Enum => (Enum, "enum"),
    EnumArray => (TextArray, "string-array"),
    Bytes => (Bytes, "bytes"),
    Boolean => (Boolean, "bool"),
    Char => (Char, "char"),
    Numeric => (Numeric, "decimal"),
    Json => (Json, "json"),
    Xml => (Xml, "xml"),
    Uuid => (Uuid, "uuid"),
    DateTime => ( DateTime, "datetime"),
    Date => (Date, "date"),
    Time => (Time, "time");

    Int32[] => (Int32Array, "int-array"),
    Int64[] => (Int64Array, "int64-array"),
    Float[] => (FloatArray, "float-array"),
    Double[] => (DoubleArray, "double-array"),
    Text[] => (TextArray, "string-array"),
    Enum[] => (TextArray, "string-array"),
    Bytes[] => (BytesArray, "bytes-array"),
    Boolean[] => (BooleanArray, "boolean-array"),
    Char[] => (CharArray, "char-array"),
    Numeric[] => (NumericArray, "numeric-array"),
    Json[] => (JsonArray, "json-array"),
    Xml[] => (XmlArray, "xml-array"),
    Uuid[] => (UuidArray, "uuid-array"),
    DateTime[] => (DateTimeArray, "datetime-array"),
    Date[] => (DateArray, "date-array"),
    Time[] => (TimeArray, "time-array");

    (Unknown, "unknown"),
    (UnknownArray, "unknown-array")
}
```

Didn't work well because `EnumArray` maps to `TextArray`, but so does `Text[]`. Also, I need to map `Array[]` and `EnumArray[]` to `unreachable!()`. Lemme know if you find a better way.

Reply:
I think the problem can be solved through procedural macro, because you can de-duplicate enum variants. This, however, is a bit more effort. I will leave it up to you to decide which approach to go with, I don't feel strongly about it